lib/model/unreads.dart

```diff
@@ -0,0 +1,383 @@
+        if (event.messageIds.length == 1) {
+          _removeInStreamTopic(event.messageIds.single, streamId, topic);
+        } else {
+          // Could construct [Set] at the edge as a field on the event,
+          // if other consumers want it (but see TODO in _removeAllInStreamTopic)
+          _removeAllInStreamTopic(Set.of(event.messageIds), streamId, topic);
+        }
+      case MessageType.private:
+        _slowRemoveAllInDms(event.messageIds);
+    }
+    notifyListeners();
+  }
+
+  void handleUpdateMessageFlagsEvent(UpdateMessageFlagsEvent event) {
+    switch (event.flag) {
+      case MessageFlag.starred:
+      case MessageFlag.collapsed:
+      case MessageFlag.hasAlertWord:
+      case MessageFlag.historical:
+      case MessageFlag.unknown:
+        // These are irrelevant.
+        return;
+
+      case MessageFlag.mentioned:
+      case MessageFlag.wildcardMentioned:
+        // Empirically, we don't seem to get these events when a message is edited
+        // to add/remove an @-mention, even though @-mention state is represented
+        // as flags. Instead, we just get the [UpdateMessageEvent], and that
+        // contains the new set of flags, which we'll use to update [mentions].
+        // (See our handling of [UpdateMessageEvent].)
+        //
+        // Handle the event anyway, using the meaning on the tin.
+        // It might be used in a valid case we haven't thought of yet.
+        switch (event) {
+          case UpdateMessageFlagsAddEvent():
+            mentions.addAll(
+              event.messages.where(
+                (messageId) => _slowIsPresentInStreams(messageId) || _slowIsPresentInDms(messageId),
+              ),
+            );
+
+          case UpdateMessageFlagsRemoveEvent():
+            mentions.removeAll(event.messages);
+        }
+
+      case MessageFlag.read:
+        switch (event) {
+          case UpdateMessageFlagsAddEvent():
+            if (event.all) {
+              streams.clear();
+              dms.clear();
+              mentions.clear();
+              oldUnreadsMissing = false;
+            } else {
+              mentions.removeAll(event.messages);
+              _slowRemoveAllInStreams(event.messages);
+              _slowRemoveAllInDms(event.messages);
+            }
+          case UpdateMessageFlagsRemoveEvent():
+            final newlyUnreadInStreams = <int, Map<String, QueueList<int>>>{};
+            final newlyUnreadInDms = <DmNarrow, QueueList<int>>{};
+            for (final messageId in event.messages) {
+              final detail = event.messageDetails![messageId];
+              if (detail == null) { // TODO(log) if on Zulip 6.0+
+                // Happens as a bug in some cases before fixed in Zulip 6.0:
+                //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.20in.20unsubscribed.20streams/near/1458467
+                // TODO(server-6) remove Zulip 6.0 comment
+                return;
```

Comment:
Perhaps better just `continue` — if we run into a bug like this, seems best to carry on with any messages we do have the details for.